hpx/parallel/segmented_algorithms/detail/scan.hpp

```diff
@@ -0,0 +1,792 @@
+
+            segment_iterator sit = traits::segment(first);
+            segment_iterator send = traits::segment(last);
+
+            typedef typename std::vector<T> vector_type;
+
+            std::vector<vector_type> results;
+
+            // scan on each partition, push whole result vector into results
+            if (sit == send)
+            {
+                // all elements on the same partition
+                local_iterator_type beg = traits::local(first);
+                local_iterator_type end = traits::end(sit);
+                if (beg != end)
+                {
+                    results.push_back(dispatch(traits::get_id(sit),
+                        Algo(), policy, std::true_type(), beg, end, op));
+                }
+            }
+            else
+            {
+                // handle the remaining part of the first partition
+                local_iterator_type beg = traits::local(first);
+                local_iterator_type end = traits::end(sit);
+
+                if (beg != end)
+                {
+                     results.push_back(dispatch(traits::get_id(sit),
+                        Algo(), policy, std::true_type(), beg, end, op));
+                }
+
+                // handle all partitions
+                for(++sit; sit != send; ++sit) {
+                    beg = traits::begin(sit);
+                    end = traits::end(sit);
+                    if (beg != end)
+                    {
+                        results.push_back(dispatch(traits::get_id(sit),
+                            Algo(), policy, std::true_type(), beg, end, op));
+                    }
+                }
+
+                // handle the beginning of the last partition
+                beg = traits::begin(sit);
+                end = traits::local(last);
+                if (beg != end)
+                {
+                    results.push_back(dispatch(traits::get_id(sit),
+                        Algo(), policy, std::true_type(), beg, end, op));
+                }
+            }
+
+            // merge results with given merge algorithm f1
+            // update init value with function f2
+            T last_value = init;
+            for (auto res : results) {
+                dest = f1(res.begin(), res.end(),
+                    dest, last_value, op);
+                last_value = f2(res, last_value);
+            }
+
+            return result::get(std::move(dest));
+        }
+
+        ///////////////////////////////////////////////////////////////////////
+        // parallel implementation
+
+        // parallel segmented OutIter implementation
+        template <typename Algo, typename ExPolicy, typename SegIter, typename OutIter,
+            typename T, typename Op>
+        static typename util::detail::algorithm_result<ExPolicy, OutIter>::type
+        segmented_scan_par(ExPolicy const& policy, SegIter first,
+            SegIter last, OutIter dest, T init, Op && op, std::true_type)
```

Comment:
Once `init` has been copied at the uppermost API level it could be passed down by `const&`.